### PR TITLE
Sentiment UI

### PR DESCRIPTION
**Pull request by HotLemonJuice**
To show the sentiment score on the map.

Details:
- Use random number for 0 - 4 to fake sentiment score for now
- Use sentiment score as the `count` for each polygon
- Disable the normalization toggle and set it to off when the sentiment toggle is on because it makes no sense to normalize sentiment scores
screen shot 2017-04-26 at 15 20 06

### CODE REVIEW

File: neo/public/javascripts/map/controllers.js [hunk 1]
```diff
@@ -371,19 +373,31 @@ angular.module('cloudberry.map', ['leaflet-directive', 'cloudberry.common'])
...
+        if($scope.doSentiment){  // 0 <= d <= 4
+          if( d < 1 * 4 / 3) {
+            return sentimentColors[0];
+          } else if( d < 2 * 4 / 3){
+            return sentimentColors[1];
+          } else{
+            return sentimentColors[2];
+          }
+        }
+        else{
```
Comment (JavierJia) [inline neo/public/javascripts/map/controllers.js]:
```
} else {
```

File: neo/public/javascripts/map/controllers.js [hunk 2]
```diff
@@ -371,19 +373,31 @@ angular.module('cloudberry.map', ['leaflet-directive', 'cloudberry.common'])
...
       function getColor(d) {
-        if(!d || d <= 0) {
-          d = 0;
-        } else if (d ===1 ){
-          d = 1;
-        } else {
-          d = Math.ceil(Math.log10(d));
-          if(d <= 0) // treat smaller counts the same as 0
-            d = 0
+        if($scope.doSentiment){  // 0 <= d <= 4
```
Comment (JavierJia) [inline neo/public/javascripts/map/controllers.js]: could you make two functions `getSentimentColor` and `getCountColor`, then make this `getColor` more simpler?

File: neo/public/javascripts/map/controllers.js [hunk 3]
```diff
@@ -444,10 +458,17 @@ angular.module('cloudberry.map', ['leaflet-directive', 'cloudberry.common'])
...
-                if ($scope.doNormalization)
-                  setNormalizedCount(geo, r);
-                else
-                  setUnnormalizedCount(geo, r);
+                if($scope.doSentiment){
+                  // TODO: change fake random number to real sentiment (0-4)
+                  geo['properties']['count'] = Math.random() * 4;
+                  geo["properties"]["countText"] = geo["properties"]["count"].toFixed(1);
+                }
+                else{
```
Comment (JavierJia) [inline neo/public/javascripts/map/controllers.js]: `} else {`

File: neo/public/javascripts/map/controllers.js [hunk 4]
```diff
@@ -471,44 +496,56 @@ angular.module('cloudberry.map', ['leaflet-directive', 'cloudberry.common'])
...
+        if($scope.doSentiment){
+          div.setAttribute("title", "Sentiment Score: Negative(0)-Positive(4)");  // add tool-tips for the legend
+          div.innerHTML +=
+            '<i style="background:' + getColor(1) + '"></i>Negative<br>';
+          div.innerHTML +=
+            '<i style="background:' + getColor(2) + '"></i>Neutral<br>';
+          div.innerHTML +=
+            '<i style="background:' + getColor(3) + '"></i>Positive<br>';
         }
+        else{
```
Comment (JavierJia) [inline neo/public/javascripts/map/controllers.js]: `} else {`

**Comment by chenlica** [comment]
Since this PR is adding a new "face" to the UI, it's time to think about how to allow the users to choose the different faces.  One possible design is to provide a few options such as "Density, Normalized Density, Sentiment".  Based on the choice, we show the corresponding face.  Thoughts?

**Comment by HotLemonJuice** [comment]
@JavierJia I have addressed your comments and refactored the code a lot, especially in the legend/normalizeToggle/sentimentToggle part. Can you review the new changes?

**Comment by HotLemonJuice** [comment]
@chenlica Currently, we use toggles to let users switch between faces. However, I found the sidebar from the MapD example could be a better option to list different options for the UI for the following reasons:
(1) It takes less space
(2) We can potentially list more options than stacking them 
screen shot 2017-04-27 at 10 52 42
screen shot 2017-04-27 at 10 52 50

**Comment by chenlica** [comment]
@HotLemonJuice We can use something similar to their UI.  In general, we expect to add various analysis functionalities to the UI, so it's important to allow the user to select different "faces."

File: neo/public/javascripts/map/controllers.js [hunk 5]
```diff
@@ -71,7 +72,8 @@ angular.module('cloudberry.map', ['leaflet-directive', 'cloudberry.common'])
           dashArray: '',
           fillOpacity: 0.7
         },
-        colors: [ '#f7f7f7', '#92c5de', '#4393c3', '#2166ac', '#f4a582', '#d6604d', '#b2182b']
+        colors: [ '#ffffff', '#92c5de', '#4393c3', '#2166ac', '#f4a582', '#d6604d', '#b2182b'],
+        sentimentColors: ['#ff0000', '#C0C0C0', '#00ff00']
```
Comment (JavierJia) [inline neo/public/javascripts/map/controllers.js]: 👍

File: neo/public/javascripts/map/controllers.js [hunk 6]
```diff
@@ -444,10 +464,16 @@ angular.module('cloudberry.map', ['leaflet-directive', 'cloudberry.common'])
...
             angular.forEach(result, function (r) {
               if (r[level] === geo['properties'][level+"ID"]){
-                if ($scope.doNormalization)
-                  setNormalizedCount(geo, r);
-                else
-                  setUnnormalizedCount(geo, r);
+                if($scope.doSentiment){
+                  // TODO: change fake random number to real sentiment (0-4)
+                  geo['properties']['count'] = Math.random() * 4;
+                  geo["properties"]["countText"] = geo["properties"]["count"].toFixed(1);
```
Comment (JavierJia) [inline neo/public/javascripts/map/controllers.js]: should we regulate the string format? do you know which is norm? `"` or `'`?

**Comment by chenlica** [comment]
@JavierJia : OK, let's merge this one, then do the improvements step by step.

File: neo/public/javascripts/map/controllers.js (hunk 6 above)
Comment (HotLemonJuice) [inline neo/public/javascripts/map/controllers.js]: As often indicated by the codacy, double quote is the norm to follow. There are many places in the project involving double quotes and single quotes. I think we will need a dedicated PR to fix it.

**Pull request merged**